packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.swift

```diff
@@ -253,16 +253,23 @@ public class InAppPurchasePlugin: NSObject, FlutterPlugin, FIAInAppPurchaseAPI {
     let pendingTransactions = getPaymentQueueHandler().getUnfinishedTransactions()
 
     for transaction in pendingTransactions {
+      // finishTransaction() cannot be called on a Transaction with a current purchasing state
+      // https://developer.apple.com/documentation/storekit/skpaymentqueue/1506003-finishtransaction
+      guard transaction.transactionState != SKPaymentTransactionState.purchasing else {
+        return
+      }
+
       // If the user cancels the purchase dialog we won't have a transactionIdentifier.
       // So if it is null AND a transaction in the pendingTransactions list has
       // also a null transactionIdentifier we check for equal product identifiers.
-      if transaction.transactionIdentifier == transactionIdentifier
-        || (transactionIdentifier == nil
-          && transaction.transactionIdentifier == nil
-          && transaction.payment.productIdentifier == productIdentifier)
-      {
-        getPaymentQueueHandler().finish(transaction)
+      guard
+        transaction.transactionIdentifier == transactionIdentifier
+          || (transactionIdentifier == nil && transaction.transactionIdentifier == nil
+            && transaction.payment.productIdentifier == productIdentifier)
```

Comment:
if we can't separate this, can we assign a meaningful name to it?
e.g. 
```
let isCanceled = transactionIdentifier == nil
```

Reply:
oh yes i can do that